bin/nvidia-container-toolkit

```diff
@@ -2,12 +2,38 @@
 
 set -eu
 
-# Ensure nvidia support setup correctly #
+device_wait() {
+
+    COUNT=0
+    SLEEP=3
+    TRIES=10
+
+    echo "Waiting for device to become available: ${1}"
+
+    while [ ${COUNT} -le ${TRIES} ] ; do
+        echo "Checking device: ${COUNT}/${TRIES}"
+        test -c "${1}"
+        if [ $? -eq 0 ] ; then
+            echo "Device found"
+            return 0
+        fi
+        sleep $SLEEP
+        COUNT=$(($COUNT + 1))
+    done
+
+    echo "Device not found"
+    return 1
+
+}
+
+# Ensure nvidia support setup correctly, and only if hardware is preset and correct #
 if snapctl is-connected graphics-core22 ; then
 
-    # Only set this up if the system has an nvidia card available #
+    # If module is loaded - hardware _should_ be present #
     grep -Eq "^nvidia" /proc/modules || exit 0
```

Comment:
this is not needed and racy, as module may not be loaded yet, whilst the device_wait function would wait for that to happen.

Reply:
The main reason I left this in was because it seemed like a nice way of simply skipping most of the logic if there was no chance of the hardware being present at all.  But, perhaps as you point out, out that assumption could be flawed.

Without this, the `dockerd` service would always wait for the `nvidia-container-toolkit` to finish [ 30 secs with current wait log timing settings ] on systems that simply don't have nvidia hardware.

Given the intent here is to not artificially delay setup on a system with no hope of an nvidia device appearing, can you think of any alternative/better way that would not require any overly privileged snap interface ?

Reply:
1. device needs to be present (`lspci -d 10de:` should be non-empty)
2. On classic - kernel module needs to be loaded & /hostfs libraries should be available and usable
3. On core - nvidia-assemble needs to be installed and running (i think some other snaps export a socket over an interface, or just export an empty interface, to allow others to connect and thus observe that nvidia-assemble is installed/running).

However, I hope that graphics-core22 is not automatically installed and connected, thus maybe installing and connecting that should be enough as a wait?

Can ponder and improve that later.

Reply:
@xnox 

I have been thinking abou tthis, and was just about to commit another change to make the NVIDIA support "opt in" only, via snap config.  I think that address most concerns for now.

Shall I add it and see what you think ?

Reply:
not sure. it still feels like an opt-out to me. If you have hardware, and drivers, and connect them up, one likely wants to use them.

Reply:
ok, cool, that work for me - I will propose an opt-out approach in a separate PR :+1: